### PR TITLE
CSS immediate children (:scope selector)

### PR DESCRIPTION
**Pull request by sortafreel**
CSS immediate children pseudo-element (from CSS `:scope > div` to XPath `child::div` or `./div`). Works only at the start of a selector. https://developer.mozilla.org/en-US/docs/Web/CSS/:scope 

Needed to get immediate children of a processed selector in Scrapy:
```
for product in response.css('.product'):
    description = product.css(':scope > div::text').get()
```
HTML example:
```
<body>
    <div class="product">
        <p class="price">
            <div>$99.99</div>
        </p>
        <div>
            Some description text
        </div>
        <p>
            <div>Shipping info</div>
        </p>
    </div>
    <div class="product">
        <p class="price">
            <div>$199.99</div>
        </p>
        <div>
            Another description text
        </div>
        <p>
            <div>Another shipping info</div>
        </p>
    </div>
</body>
```

### CODE REVIEW

**Comment by sortafreel** [comment]
Testing, adding custom tests.

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,6 @@
 /dist
 /docs/_build
 /.coverage
-.idea
\ No newline at end of file
+.idea
+/venv
+*.vscode
```
Comment (dangra) [inline .gitignore]: Keep editor preferences local to your setup (that is your $HOME/.gitignore), do not include them in the project.

File: cssselect/xpath.py [hunk 2]
```diff
@@ -220,7 +220,18 @@ def selector_to_xpath(self, selector, prefix='descendant-or-self::',
...
             xpath = self.xpath_pseudo_element(xpath, selector.pseudo_element)
-        return (prefix or '') + _unicode(xpath)
+
+        unicode_xpath = _unicode(xpath)
+        # CSS immediate children (CSS "<> div" to XPath "child::div" or "./div")
+        # Works only at the start of a selector
+        # Needed to get immediate children of a processed selector in Scrapy
+        # product = response.css('.product')
+        # name = product.css('<> div')
+        child_re = r'^[ \t\r\n\f]*\<[ \t\r\n\f]*\/'
```
Comment (dangra) [inline cssselect/xpath.py]: precompile the pattern outside of the hot path. A class level attribute is fine.

File: cssselect/xpath.py [hunk 3]
```diff
@@ -332,6 +343,9 @@ def xpath_element(self, selector):
         if not element:
             element = '*'
             safe = True
+        if element == '<':
+            element = '<'
```
Comment (dangra) [inline cssselect/xpath.py]: why the reassignment to `<` ?

File: .gitignore (hunk 1 above)
Comment (sortafreel) [reply .gitignore]: ACK :)

File: cssselect/xpath.py (hunk 2 above)
Comment (sortafreel) [reply cssselect/xpath.py]: Understood, thanks :)

File: cssselect/xpath.py (hunk 3 above)
Comment (sortafreel) [reply cssselect/xpath.py]: Previously was reassigning to '.', will fix :)

Comment (sortafreel) [reply cssselect/xpath.py]: Fixed.

File: cssselect/xpath.py (hunk 2 above)
Comment (sortafreel) [reply cssselect/xpath.py]: Fixed.

File: .gitignore (hunk 1 above)
Comment (sortafreel) [reply .gitignore]: Fixed.

**Comment by alexander-matsievsky** [comment]
@sortafreel Great job, Alexander!

One thought though. The [Selectors Level 4](https://www.w3.org/TR/selectors-4/#scope-element) defines the `:scope` pseudo-class. From https://developer.mozilla.org/en-US/docs/Web/CSS/:scope:

> When used from a DOM API such as querySelector(), querySelectorAll(), matches(), or Element.closest(), :scope matches the element you called the method on.

Could we use it instead of `<` (or other non-standard alternatives like `@`, `&`, etc.)? The reasoning is to avoid extending CSS in an ad-hoc manner and to save on the refactoring when [cssselect](https://cssselect.readthedocs.io/en/latest/#supported-selectors) supports CSS4 selectors eventually.

File: cssselect/xpath.py (hunk 2 above)
Comment (sortafreel) [reply cssselect/xpath.py]: Fixed :)

**Comment by sortafreel** [comment]
> I love these changes! The only thing I think is missing is to update https://cssselect.readthedocs.io/en/latest/#supported-selectors to mention `:scope`.

Done :)

**Comment by dangra** [comment]
@sortafreel seems we are very close, only rest covering the case codecov is complaining with a test.

**Comment by Gallaecio** [comment]
@dangra The tests cover all changes as far as @sortafreel and me can tell. @sortafreel tried different changes to see if they had any effect on codecov, but neither @sortafreel nor me managed to find a new path of code that is not hit by at least one test.

**Comment by Gallaecio** [comment]
I believe we are being hit by https://bugs.python.org/issue2506

If I do what https://bitbucket.org/ned/coveragepy/issues/594/problems-with-branch-identification-and#comment-38845397 suggests:

```
            if stream.peek() != ('DELIM', '('):
                    # …
                if result.__repr__() == 'Pseudo[Element[*]:scope]':
                    # …
                # 2 lines added here:
                else:
                    a = 1  # https://bitbucket.org/ned/coveragepy/issues/594/problems-with-branch-identification-and#comment-38845397
                continue
```

Then the coverage data from `tox -e py36` (after adding `show_missing = True` under `[report]` at `.coveragerc` to show lines missing in the coverage) no longer shows those lines as not covered.

**Comment by sortafreel** [comment]
@Gallaecio thanks for the support :)

**Comment by dangra** [comment]
Nice catch @Gallaecio !  all good then.

**Comment by dangra** [comment]
@sortafreel thanks for all the good work

**Pull request merged**